### ISSUE TITLE
[ENHANCEMENT]: Clean up the `attempt_insert_stable` logic

### ISSUE THREAD

**Issue by PointKernel** [NVIDIA/cuCollections#530]
### Is your feature request related to a problem? Please describe.

The `attempt_insert_stable` function calls `packed_cas` or `cas_dependent_write` depending on the slot type but does not ensure stable insertion. This is incorrect.

Currently, we call `wait_for_payload` before returning the device `insert_and_find` to ensure the payload is in place. This operation is supposed to occur before `attempt_insert_stable` returns.

### Describe the solution you'd like

Remove all existing invocations of `wait_for_payload` and update `attempt_insert_stable` as:
```cuda
  template <typename Value>
  [[nodiscard]] __device__ insert_result attempt_insert_stable(value_type* address,
                                                               value_type const& expected,
                                                               Value const& desired) noexcept
  {
    if constexpr (sizeof(value_type) <= 8) {
      return packed_cas(address, expected, desired);
    } else {
      auto const res = cas_dependent_write(address, expected, desired);
      if constexpr (has_payload) {
        // wait to ensure that the write to the value part also took place
        this->wait_for_payload(address->second, this->empty_value_sentinel());
      }
      return res;
    }
  }
```

The above sample fails tests for `cas_dependent_write` cases but it shows the idea.

By doing this, we can avoid calling `wait_for_payload` repeatedly, making `attempt_insert_stable` the only place where waiting occurs.

**Comment by sleeepyjack** [comment]
Shall we rename it to `attempt_insert_sync` then since the sync on the payload (or really the entire slot) is the crucial difference here?

**Comment by PointKernel** [comment]
The `_sync` suffix seems a bit misleading since it's not about synchronization, e.g. https://github.com/NVIDIA/cuCollections/blob/a547e8f8f2269daeecb8f37c22ab641c330d863d/include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh#L616-L622

Even if it syncs at line 622, the `wait_for_payload` is still mandatory to ensure the payload is updated. Additionally, the C++ convention generally names all asynchronous APIs with the `_async` suffix, rather than adding a `_sync` suffix for synchronous ones.

**Comment by srinivasyadav18** [comment]
> Remove all existing invocations of wait_for_payload and update attempt_insert_stable

`wait_for_payload` is not only used with `attempt_insert_stable`, but in many other instances (for e.g. https://github.com/NVIDIA/cuCollections/blob/687948a8ca69e6ef06334561e1f0ead97e7f19c0/include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh#L515).

So it would be not possible to remove all the instances of `wait_for_payload`, but only the instances which are associated with `attempt_insert_stable` right ?

**Comment by PointKernel** [comment]
> wait_for_payload is not only used with attempt_insert_stable, but in many other instances (for e.g.

@srinivasyadav18 If we wait at the end of `attempt_insert_stable`, the invocation you mentioned at line 515 becomes redundant, right?

**Comment by sleeepyjack** [comment]
Not sure if my train of thoughts is correct: Let's say thread `A` swaps a new pair `(k, v1)` into a slot using `attempt_insert_stable`. After this function returns (with success), both `k` and `v1` is visible to `A`. However, a concurrent thread `B` that probes for `(k, v2)` might see the key already been materialized while the dependent write of `v1` from `A` is still in-flight. Thus `B` has to wait for the payload too before it can work with the value.

In other words: There might be calls to `wait_for_payload` outside of `attempt_insert_stable`.

**Comment by PointKernel** [comment]
> Not sure if my train of thoughts is correct: Let's say thread `A` swaps a new pair `(k, v1)` into a slot using `attempt_insert_stable`. After this function returns (with success), both `k` and `v1` is visible to `A`. However, a concurrent thread `B` that probes for `(k, v2)` might see the key already been materialized while the dependent write of `v1` from `A` is still in-flight. Thus `B` has to wait for the payload too before it can work with the value.
> 
> In other words: There might be calls to `wait_for_payload` outside of `attempt_insert_stable`.

You are right, waiting within thread `A` **cound not** guarantee the memory order for thread `B` (or any other threads) so this cleanup doesn't make any sense.

Closing this issue.
